### ISSUE TITLE
Example bazel build fails due to "unexpected short read from file"

### ISSUE THREAD

**Issue by col6y** [google/bazel#174]
I compiled and installed Bazel on a 64-bit VM running 32-bit Debian 8.0.

    $ uname -a
    Linux devel-libca-skeggsc 3.16.0-4-686-pae #1 SMP Debian 3.16.7-ckt9-3~deb8u1 (2015-04-24) i686 GNU/Linux

When I create a new workspace and try to build in it, it fails with a very odd error. To make sure that it wasn't an artifact of the workspace, I rebooted the virtual machine, created a new workspace, removed `~/.cache/bazel`, and ran an example build command in the new workspace:

    skeggsc@devel-libca-skeggsc:~/new-workspace$ bazel build examples/java-native/src/main/java/com/example/myproject:hello-world
    Extracting Bazel installation...
    ..................................
    ERROR: no such package 'external': Error encountered while dealing with the WORKSPACE file: Unexpected short read from file '/home/skeggsc/.cache/bazel/_bazel_skeggsc/install/c856fa2026bd5b1d370edcd52403a9ea/_embedded_binaries/jdk.WORKSPACE' (expected 8800387990324, got 820 bytes).
    INFO: Elapsed time: 1.896s

This is the error that I've been getting from Bazel.

I checked the size of the named file:

    skeggsc@devel-libca-skeggsc:~/new-workspace$ wc -c /home/skeggsc/.cache/bazel/_bazel_skeggsc/install/c856fa2026bd5b1d370edcd52403a9ea/_embedded_binaries/jdk.WORKSPACE
    820 /home/skeggsc/.cache/bazel/_bazel_skeggsc/install/c856fa2026bd5b1d370edcd52403a9ea/_embedded_binaries/jdk.WORKSPACE

I also notice that 8800387990324 (the length that it thinks the file should be) is equal to `2^43 + 2^32 + 820`. This seems suspicious.

Any idea why I'm getting this error?

**Comment by ademuri** [comment]
I'm also seeing this error, albeit with a slightly different expected size:

```
~/bazel$ bazel build //scripts:bazel-complete.bash
.................................................................................................
ERROR: no such package 'external': Error encountered while dealing with the WORKSPACE file: Unexpected short read from file '/home/adam/.cache/bazel/_bazel_adam/install/110db003897d947b0a1eb67b57393229/_embedded_binaries/jdk.WORKSPACE' (expected 8813272892212, got 820 bytes).
INFO: Elapsed time: 2.733s
```

I'm on Ubuntu 12.04 32-bit. I installed the Oracle JDK 8 as specified at http://bazel.io/docs/install.html

This happens when trying to build either the bazel bash completion script or the same example as OP.

**Comment by kchodorow** [comment]
Reproduced! Yeah, looks like some kind of overflow, I'll look into it.

**Comment by col6y** [comment]
Confirmed - bug is fixed for me. Thanks!

**Comment by kchodorow** [comment]
Great, glad to hear it!

**Comment by benley** [comment]
I'm seeing this again on Darwin, or perhaps something very similar to it:

```
ERROR: /private/var/tmp/_bazel_bstaffin/8e8afcd0b7aa33ffce54d61976a078fa/external/prometheus-client-golang/BUILD:1: Unexpected short read from file '/private/var/tmp/_bazel_bstaffin/8e8afcd0b7aa33ffce54d61976a078fa/external/prometheus-client-golang/BUILD' (expected 169, got 121 bytes).
```

The file in question is indeed 121 bytes long.  I've seen this a few times tonight; at first I thought it was related to NFS somehow, so I took NFS completely out of the equation and it's still happening.

**Comment by col6y** [comment]
@benley that's probably a different bug - this one is related to integer overflow.

**Comment by kchodorow** [comment]
@benley this is a known issue and a byproduct of external repositories being assumed immutable. Blaze caches the size of the BUILD file but also does a sanity check that the current stat of the file matches the cached size (which of course, it does not).

We're working on removing this assumption of immutability, which should fix the issue (#352 is the bug for that).  

https://github.com/bazelbuild/bazel/commit/4e69824ba0d0b56e5426ec89ab3bd9e05f761cb5 hit this, you can see @haxorz changed src/test/shell/bazel/local_repository_test.sh so that the BUILD file modifications left the BUILD file size the same.
